# Getting started

For this workflow, we use the following software:

1. [**prodigal**](https://github.com/hyattpd/Prodigal) v2.6.3
1. [**usearch**](https://www.drive5.com/usearch/) v9.0.2132 (64 bit)
1. [**diamond**](https://github.com/bbuchfink/diamond/) v0.9.22
1. [**hmmer**](http://hmmer.org/) v3.1b2
1. [**signalp**](https://www.ncbi.nlm.nih.gov/pubmed/28451972) v4.1

----

### Gene prediction

To begin annotation, **prodigal** is used to create predictions for both the nucleotide and protein sequences of predicted ORFs in your genome. One
downside of **prodigal** is that it has a habit of dumping a lot of metadata into your fasta sequence header, which can confuse the downstream annotation
tools. **Prodigal** can be run on *NeSI*, with appropriate header correction, using the commands

```
prodigal -q -p single -i genome.fna -d genome.prod.fna -a genome.prod.faa

sed 's/ .*//g' genome.prod.faa > genome.prod.no_metadata.faa
sed 's/ .*//g' genome.prod.fna > genome.prod.no_metadata.fna
```

This takes the input file genome.fna and detects genes along each contig which are written out to fasta files in nucleotide (**-d**) and amino acid (**-a**) space.

*Note: **prodigal** has two modes of running, one for metagenomic data (**-p meta**) and one for genome-resolved data (**-p single**). The documentation recommends using single mode for genome data and while the majority of called genes are unaffected by the mode, differences can be seen between the outputs. Annecdotally, metagenomic mode will identify slightly fewer genes than single mode for a genome, but single mode can miss laterally transfered elements. There is not neccesarily a best choice for which version to use and this is at the users discretion.*

----

### Protein annotation

There are a number of ways to annotate protein data in this pipeline, based either on sequence alignment or profiling motifs according to hidden Markov models (HMMs). Alignment-based annotation is based on the traditional **BLAST** software, although **BLAST** itself is rarely used in metagenomics due to its slow run time. Two rapid alternatives to **BLAST** are **usearch** and **diamond**, for which a number of databases are precompiled on *NeSI* and provided as variables under the **annotation_pipe** command.

#### BLAST-like annotation

While **usearch** and **diamond** work in different ways, the following commands will produce output as a tab-delimited file with the following columns:

1. Query sequence
1. Target sequence
1. Alignment start position
1. Alignment stop position
1. Bitscore
1. E-value

**Usearch** will utilise all threads available whereas **diamond** uses a default of 1, with the user specifying more if desired. When writing your *slurm* script, make sure to set the number of threads to the number you intend to use.

To annotate your gene predictions against *UniProt*, *UniRef100*, and *KEGG* using **usearch**, the following commands are used

```
usearch9.0.2132_i86linux64 -id 0.5 -evalue 0.001 -maxhits 10 -top_hits_only \
                           -userfields query+target+tlo+thi+id+bits+evalue \
                           -db $UNIPROT_udb -usearch_global genome.prod.no_metadata.faa -userout genome.uniprot.txt

usearch9.0.2132_i86linux64 -id 0.5 -evalue 0.001 -maxhits 10 -top_hits_only \
                           -userfields query+target+tlo+thi+id+bits+evalue \
                           -db $UNIREF100_udb -usearch_global genome.prod.no_metadata.faa -userout genome.uniref100.txt

usearch9.0.2132_i86linux64 -id 0.5 -evalue 0.001 -maxhits 10 -top_hits_only \
                           -userfields query+target+tlo+thi+id+bits+evalue \
                           -db $KEGG_udb -usearch_global genome.prod.no_metadata.faa -userout genome.kegg100.txt
```

*Note: There are two ways for running **usearch** alignments, either with the **-usearch_global** or **-usearch_local** alignment parameter. For a first-pass at the data, I prefer to use **-usearch_global** since it has the effect of consider the database 'true', and testing how well your predicted proteins match against it. A local alignment will perform a more traditional 2-way alignment between query and target, and this will most likely give you more hits.*

To achieve equivalent outputs using **diamond** with 16 threads, use the commands

```
diamond blastp -p 16 -k 10 --quiet --outfmt 6 qseqid sseqid sstart send pident bitscore evalue \
               -d $UNIPROT_dmnd -q genome.prod.no_metadata.faa -o genome.uniprot.txt

diamond blastp -p 16 -k 10 --quiet --outfmt 6 qseqid sseqid sstart send pident bitscore evalue \
               -d $UNIREF100_dmnd -q genome.prod.no_metadata.faa -o genome.uniref100.txt

diamond blastp -p 16 -k 10 --quiet --outfmt 6 qseqid sseqid sstart send pident bitscore evalue \
               -d $KEGG_dmnd -q genome.prod.no_metadata.faa -o genome.kegg.txt
```

#### HMM-based annotation

An alternative to alignment-based annotation is to use a statistical model (HMM) to identify conserved motifs across your gene. This can be useful when trying to identify or annotate catalytic domains, or when you have short proteins or proteins that donâ€™t annotate well across the full alignment. The standard software for performing this kind of annotation is **hmmer**, and we have three **hmmer**-compatible databases available on *NeSI*. To annotate your genome against all three databases, use the command

```
hmmsearch --tblout genome.pfam.txt -E 1e-3 --cpu 10 $PFAM_hmm genome.prod.no_metadata.faa

hmmsearch --tblout genome.tigr.txt -E 1e-3 --cpu 10 $TIGRFAM_hmm genome.prod.no_metadata.faa

hmmsearch --tblout genome.eggnog.txt -E 1e-3 --cpu 10 $EGGNOG_hmm genome.prod.no_metadata.faa
```

#### Signal peptides

As an extension to gene annotation, if it often valuable to look at the signal peptides attached to gene predictions to make determinations as to whether the proteins are exported from the organisms cell or embedded in the membrane (for example, signalling receptors, and some respiratory proteins).

Performing this prediction requires knowledge as to whether your genome is from a Gram negative or positive organism (if prokaryotic) or if it is from a eukaroytic organism. If you have a prokaryotic genome, the cell wall struture (Gram straining) can be deduced from the phylum to which your genome is classified.

Performing the annotation in each way is given below, but for each bin, only perform a single step.

```
signalp -f short -s best -c 0 -t gram- genome.prod.no_metadata.faa > genome.signalp

signalp -f short -s best -c 0 -t gram+ genome.prod.no_metadata.faa > genome.signalp

signalp -f short -s best -c 0 -t euk genome.prod.no_metadata.faa > genome.signalp
```

----