# Getting started

*Note: This workflow is current being replaced by a more comprehensive tool to create a unified database of genomic, bin, taxonomy, and annotation information. While it is still useful, this tool is no longer being developed. A more powerful and flexible tool will be released soon.*

----

### Aggregating the results

Once you have obtained as many annotations are you are interested in, the files can be brought together and summarised with the **annotationAggregator.py** script. This script will take the results of all the annotations described above and produce Excel-compatible output files for you to examine. If you have not performed allt he steps above, the script is flexible in which inputs it is provided, and will just summarise whatever it is given. This means that you can re-run it as many times as desired with different input files (which might be useful if you want to start analysing your data while slower processes like **usearch** or **Infernal** run).

All options for the script can be see by invoking it with the help flag

```
annotationAggregator.py -h
```

Generally, this script will not enforce filtering criteria on the input files, with the sole exception of filtering ORF annotations obtained through **BLAST**-like searches (**usearch**, **diamond**). The default is to accept matches where the query has at least 30% amino acid similarity to the target protein, over 50% of the target length. These can be adjusted as desired, but are probably close to the lower limit of acceptable values.

To summarise the outputs of all the annotations performed in this document

```
annotationAggregator.py --ident 30.0 --coverage 70.0 \
                        -u genome.uniprot.txt -r genome.uniref100.txt -k genome.kegg.txt \
                        -s genome.signalp \
                        -p genome.pfam.txt -i genome.tigr.txt \
                        -m genome.metaxa.bacteria.fasta -t genome.aragorn \
                        -n genome.rfam \
                        genome.fna genome.prod.fna genome.prod.faa genome.annotation
```

Note that the four positional arguments at the end of the command include the raw output files from **prodigal**, not the version with the metadata removed.

This script will automatically attempt to find the other variants of the *genome.metaxa.bacteria.fasta* file. If you only want to consider the file provided to the script, use the **-â€“restrict_metaxa** flag to prevent it searching for the variant files.

The script will produce two output files, *genome.annotation.nt* and *genome.annotation.aa* which contain the nucleotide and amino acid predictions and annotations, resepectively.

----