# Getting started

For this workflow, we are using a mixture of previously generated files as well as some new data. In addition to the previous software, we are going to use:

1. [**BBMap**](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbmap-guide/) v39.73
1. [**samtools**](https://github.com/samtools/samtools) v1.9
1. [**prodigal**](https://github.com/hyattpd/Prodigal) v2.6.3
1. [**subread**](http://subread.sourceforge.net/) v1.6.3

----

#### Mapping RNA sequences to the assembly

Like with the DNA read mapping, we can map our RNA data to the assembly using **BBMap**, **bowtie**, or **bowtie2**. This is at your discretion, although we will use **BBMap** here. We then sort and compress the *sam* alignment using **samtools**.

```bash
bbmap.sh ref=contigs.fna

bbmap.sh t=30 ref=contigs.fna in1=SampleA_R1.rna.fq in2=SampleA_R2.rna.fq out=SampleA.rna.sam
bbmap.sh t=30 ref=contigs.fna in1=SampleB_R1.rna.fq in2=SampleB_R2.rna.fq out=SampleB.rna.sam

samtools sort -@ 10 -o SampleA.rna.bam SampleA.rna.sam
samtools sort -@ 10 -o SampleB.rna.bam SampleB.rna.sam
```

----

#### Creating feature table for featureCounts

We now need to create a table mapping out the coordinates of contigs which correspond to genes. In this example, we are only going to look at protein-coding sequences identified by **prodigal**, but it is informative to include tRNA and rRNA annotations as well.

Creating the table can be done easily with **python**, specifically the **pandas** library. Please note that this example is written to work under **python** version 3 and will not work for version 2.

```python
import pandas as pd
import re

# Import the prodigal file, and extract sequence names/metadata
prodigal_headers = [ line.replace('>', '') for line in open('genome.prod.faa', 'r') if '>' in line ]

# Define a function for splitting the metadata into row-wise dictionaries
def prodigal_to_dict(file_line):
    gene_name, start, stop, orientation, *rest = file_line.split('#')
    return { 'GeneID': gene_name, 
             'Chr': re.sub('_[\d+]$', '', gene_name),
             'Start': start.strip(),
             'End': stop.strip(),
             'Strand': '+' if int(orientation) == 1 else '-' }

# Convert the prodigal lines into a DataFrame
prodigal_buffer = [ prodigal_to_dict(header) for header in prodigal_headers ]
prodigal_df = pd.DataFrame(prodigal_buffer)

# Order the columns, and save the file
prodigal_df = prodigal_df[ ['GeneID', 'Chr', 'Start', 'End', 'Strand' ] ]
prodigal_df.to_csv('gene_coords.txt', sep='\t', index=False)
```

----

#### Distributing the RNA counts over gene regions

This can be done with the **featureCounts** program from the **subread** library.

```bash
featureCounts -p -T 20 -t exon -F SAF -a gene_coords.txt -o gene_counts.txt *.rna.bam
```

The results of this are stored to the file `gene_counts.txt`, which can then be imported into a program like **R** to identify the differentially expressed genes.

---
