# Recover 16S rRNA from assembled reads
Perhaps one of the most intuitive thought after obtaining metagenomic reads is "What does the community/population look like?" Often, this is investigated using 16S rRNA amplicons. You may also have done some amplicon-based survey and want to see if your analyses results from metagenomic reads are congruent with them. Unfortunately, assemblers tend to produce misassembled and fragmented  16S rRNA sequences ([Miller *et al.* 2011](http://genomebiology.com/2011/12/5/R44)). Thus, other tools are required to reconstruct and identify 16S rRNA sequences from the reads themselves. Here, we can use two tools: 
- **EMIRGE** ([GitHub page](https://github.com/csmiller/EMIRGE), [Miller *et al*. 2011](http://genomebiology.com/2011/12/5/R44), [Miller *et al*. 2013](https://doi.org/10.1371/journal.pone.0056018)) 
- **phyloFlash** ([GitHub page](http://hrgv.github.io/phyloFlash/), [Gruber-Vodicka *et al*. 2020](https://doi.org/10.1128/mSystems.00920-20)).

# EMIRGE


# phyloFlash

**phyloFlash** is a software pipeline that aims to reconstruct full-length SSU rRNA for the exploring phylogenetic composition within short-read (Illumina) (meta)genomics or (meta)transcriptomic dataset. It can use **EMIRGE** (above) and **SPAdes** to achieve this. If you have trouble getting **EMIRGE** to work on your system, **phyloFlash** may be a good alternative. Briefly, **phyloFlash** finds putative SSU rRNA reads by mapping them to a database (SILVA), filtering them for potential cloning vector sequences, then using **SPAdes** and/or **EMIRGE** to assemble the resulting reads.

## Installation

Obtain the latest release from the GitHub page and check dependencies.

```bash
# Download release
wget https://github.com/HRGV/phyloFlash/archive/pf3.4.tar.gz

# Unpack
tar -xzf pf3.4.tar.gz

# Load modules
module purge
module load Perl/5.34.1-GCC-11.3.0
module load BBMap/38.95-gimkl-2020a
module load VSEARCH/2.21.1-GCC-11.3.0
module load MAFFT/7.487-gimkl-2020a-with-extensions
module load BEDTools/2.30.0-GCC-11.3.0
module load SPAdes/3.15.4-gimkl-2022a-Python-3.10.5

# Check dependencies
cd phyloFlash-pf3.4
./phyloFlash.pl -check_env
```

Append scripts to your path

```bash
export PATH=/path/to/phyloFlash:$PATH
```

Set up the following databases:
- [UniVec](https://ftp.ncbi.nlm.nih.gov/pub/UniVec/UniVec)
- [SILVA Ref NR99](https://www.arb-silva.de/no_cache/download/archive/current/Exports/)

If you have local copies of the database

```bash
cd /path/to/phyloFlash-pf3.4

phyloFlash_makedb.pl \
  --noemirge \
  --univec_file /path/to/UniVec \
  --silva_file /path/to/SILVA
```

> **Note on `--noemirge` flag**<br>
> This prevents issues with **bowtie** used downstream. 

If you want to download and format them automatically

```bash
cd /path/to/phyloFlash-pf3.4

phyloFlash_makedb.pl --remote
```

## Run software

Here, we are using phyloFlash as an *alternative* to **EMIRGE**. By default, it will only use SPAdes to do the rRNA reconstruction.

```bash
phyloFlash.pl \
  -CPUs 16 \
  -lib sampleA \
  -read1 sampleA_R1.fastq.gz \
  -read2 sampleA_R2.fastq.gz \
  -almosteverything \
  -log
```

> **Note for transcriptomic data**<br>
> If you have transcriptomic datasets with, or tend to have, large amounts of rRNA reads, add argument `-readlimit 1000000`. 

## Compare multiple samples

If you have reads from multiple samples, you can compare all of them using the wrapper script `phyloFlash_compare.pl`.

```bash
# Load modules
module purge
module load Perl/5.34.1-GCC-11.3.0
module load R/4.2.1-gimkl-2022a

# Navigate to output directory
cd /path/to/phyloFlash_output

# Run comparison script
phyloFlash_compare.pl \
  --allzip  \
  --task barplot,heatmap,ntu_table,matrix \
  --outfmt png \
  --log
```
