# Introduction

[**DRAM**](https://github.com/WrightonLabCSU/DRAM) is a useful tool for microbial gene prediction and annotation. It also comes with a viral-focused equivalent, **DRAM-v**. **DRAM-v** outputs predicted annotations for genes based on several relevant databases, as well as predicted auxiliary metabolic genes (AMG).

To run DRAM-v we must first re-run VirSorter2 (with the filtering steps omitted this time) to prepare the necessary input files for DRAM-v (including to calculate metrics that DRAM-v uses for flagging AMG). For large data sets, it advised to partition the vOTU sequence file for parallel runs and then concatenate the results.

In the following sections, we will:
1. Split viral contigs input file
   <!--(See here for more info on parallelisation and restarting runs) See where?-->
2. Prepare AMG identification files for **DRAM-v** using **VirSorter2** (*filtering switched off*)
3. Concatenate **VirSorter2** outputs prepared for **DRAM-v** for other downstream use (taxonomy, etc.)
4. Run **DRAM-v**
5. Concatenate results

<br>

# Prepare inputs

### Split fasta file

For large data sets, split the vOTUs fna file into chunks to run **DRAM-v** in parallel.

```bash
# Load modules
module purge
module load BBMap/38.95-gimkl-2020a

# Output directories
mkdir -p viral_annotation/DRAMv/vsort2_prepfiles/split_input_fasta

partition.sh \
  in=viral_identification/checkv_vOTUs/vOTUs.checkv_filtered.fna \
  out=viral_annotation/DRAMv/vsort2_prepfiles/split_input_fasta/vOTUs_filtered_subset_%.fna \
  ways=100
```

### Re-run VirSorter2 on vOTUs

Here, the following additional arguments are required for **VirSorter2**:
- `--prep-for-dramv`
- `--viral-gene-enrich-off` 
- `--provirus-off`
- `--min-score 0`
- `--seqname-suffix-off`

See [notes in previous document](1.identify_viral_sequences.md#virsorter2) on running **VirSorter2** for NeSI users.

In this example, an array **SLURM** job with 100 files (partitioned vOTUs) are used as input.

```bash
#!/bin/bash -e
#SBATCH --account       <project_account>
#SBATCH --job-name      DRAMv_prep_VirSorter2
#SBATCH --time          01:00:00
#SBATCH --mem           2GB
#SBATCH --array         0-99
#SBATCH --cpus-per-task 24
#SBATCH --error         %x_%A_%a.err
#SBATCH --output        %x_%A_%a.out

# Load modules
module purge
module unload XALT
module load VirSorter/2.2.3-gimkl-2020a-Python-3.8.2

# Output directories
mkdir -p viral_annotation/DRAMv/vsort2_prepfiles/vOTUs_filtered_subsets/

# Run VirSorter2
virsorter run -j 24 \
  --seqname-suffix-off \
  --viral-gene-enrich-off \
  --provirus-off \
  --prep-for-dramv \
  --min-score 0 \
  --include-groups dsDNAphage,NCLDV,RNA,ssDNA,lavidaviridae \
  -i viral_annotation/DRAMv/vsort2_prepfiles/split_input_fasta/vOTUs_filtered_subset_${SLURM_ARRAY_TASK_ID}.fna \
  -d virsorter2_database/ \
  -l vOTUs_filtered_subset_${SLURM_ARRAY_TASK_ID} \
  -w viral_annotation/DRAMv/vsort2_prepfiles/vOTUs_filtered_subsets/vOTUs_filtered_subset_${SLURM_ARRAY_TASK_ID} \
  --tmpdir ${SLURM_JOB_ID}.tmp \
  --rm-tmpdir \
  all \
  --config LOCAL_SCRATCH=${TMPDIR:-/tmp}
```

### Concatenate files

```bash
cd viral_annotation/DRAMv/vsort2_prepfiles
mkdir -p vOTUs_filtered_concatenated

> vOTUs_filtered_concatenated/final-viral-combined-for-dramv.fa
for i in {0..99}; do
    cat vOTUs_filtered_subsets/vOTUs_filtered_subset_${i}/vOTUs_filtered_subset_${i}-for-dramv/final-viral-combined-for-dramv.fa \
    >> vOTUs_filtered_concatenated/final-viral-combined-for-dramv.fa
done
```

<br>

# Run DRAM-v

### Annotate

```bash
#!/bin/bash -e
#SBATCH -A your_project_account
#SBATCH -J 4_DRAMv_annotation_vOTUs
#SBATCH --time=01:00:00
#SBATCH --mem=80Gb
#SBATCH --ntasks=1
#SBATCH --array=0-99
#SBATCH --cpus-per-task=32
#SBATCH -e 4_DRAMv_annotation_vOTUs_%a.err 
#SBATCH -o 4_DRAMv_annotation_vOTUs_%a.out 

# Load modules
module purge
module load DRAM/1.3.5-Miniconda3

# Output directories
mkdir -p viral_annotation/DRAMv/dramv_annotation

# Run DRAM-v
DRAM-v.py annotate \
  --threads 32 \
  --min_contig_size 1000 \
  -i viral_annotation/DRAMv/vsort2_prepfiles/vOTUs_filtered_subsets/vOTUs_filtered_subset_${SLURM_ARRAY_TASK_ID}/vOTUs_filtered_subset_${SLURM_ARRAY_TASK_ID}-for-dramv/final-viral-combined-for-dramv.fa \
  -v viral_annotation/DRAMv/vsort2_prepfiles/vOTUs_filtered_subsets/vOTUs_filtered_subset_${SLURM_ARRAY_TASK_ID}/vOTUs_filtered_subset_${SLURM_ARRAY_TASK_ID}-for-dramv/viral-affi-contigs-for-dramv.tab \
  -o viral_annotation/DRAMv/dramv_annotation/dramv_annotation_subset_${SLURM_ARRAY_TASK_ID}
```

### Compile DRAM-v annotation subsets

The script compile_dram_annotations.py was written to recompile subsets of **DRAM** outputs, while allowing for cases where some results files were not generated (e.g. no tRNAs were identified for a given subset). It takes as input the directory path that contains each of the **DRAM-v** subsets outputs.

```bash
# Load modules
module purge
module load Python/3.8.2-gimkl-2020a

# Run compile dram annotations
compile_dram_annotations.py \
  -i viral_annotation/DRAMv/dramv_annotation \
  -o viral_annotation/DRAMv/dramv_annotation/collated_dramv_
```

### Distill

`DRAM-v.py distill` can be used to output predicted AMGs.

> **Note**<br>
> Optionally, the flags `--remove_transposons` and `--remove_fs` can be included to exclude predicted AMGs on scaffolds with transposons and those that are near the ends of scaffolds (these situations increase the likelihood of false positives).

```bash
# Load modules
module purge
module load DRAM/1.3.5-Miniconda3

# Run DRAM
DRAM-v.py distill \
  --remove_transposons \
  --remove_fs \
  -i viral_annotation/DRAMv/dramv_annotation/collated_dramv_annotations.tsv \
  -o viral_annotation/DRAMv/dramv_distillation
```

<br>

# Additional resources

[Pratama et al. (2021)](https://doi.org/10.7717/peerj.11447) is a great resource on standards in viromics, including specific discussion on AMG discovery and confirmation. 
