# Viral taxonomy

Viral taxonomy in general is a challenging and always evolving field. For example, 2022 saw a large overhaul of taxonomic idetification and ranking structures. For more information on viral taxonomy, see the [website](https://ictv.global/) for the International Committee on Taxonomy of Viruses.

In addition to this, predicting viral taxonomy remains challenging in metagenomic-based studies.

A current leading approach involves generating a gene-sharing network of the similarity of all viral genomes (or contigs) compared to reference genomes based on the extent of shared genes. [**vConTACT2**](https://bitbucket.org/MAVERICLab/vcontact2/src/master/) can generate a gene sharing network including the viral-RefSeq reference database, and which can be used to infer the potential taxonomy of your data set's viral contigs. **vConTACT2** generates this output based on predicted genes (e.g. prodigal output file). Taxonomy can then be inferred (note, this is *NOT* a definitive taxonomy assignment) based on clustering with reference genomes.

Key outputs from **vConTACT2** are:

- `genome_by_genome_overview.csv`: clustering results of all included contigs (also including ViralRefSeq genomes). In some cases, taxanomy can be inferred based on how closely your query viral genome clusters with a reference genome(s). The script `tax_predict.py` was written to automate taxonomy predictions *en masse* based on this output file. However, note that in practice this results in only a very small percentage of taxonomy predictions overall in metagenomic data sets. It is generally necessary to examine the clustering statistics and visualise the network to assess the likely taxonomy of each viral genome of interest.
- `c1.ntw`: network file that can be visualised (e.g. **Cytoscape**)

A helpful example (including visualisation in **Cytoscape**) is [available here](https://www.protocols.io/view/applying-vcontact-to-viral-sequences-and-visualizi-kqdg3pnql25z/v5?step=7)

<br>

# vConTACT2

> **Note to NeSI users**<br>
> **vConTACT2** is not currently available in NeSI. **vConTACT2** is installed as a **conda** environment. You may also need to install `cluster_one-1.0.jar` separately. It is advisable to discuss **conda** installs with a member of the NeSI support team.

Assuming a **conda** install is available, the following script does:
1. **Prodigal** gene prediction<br>
   *It may be possible to use the predicted genes from **DRAM-v**'s output as long as the format is appropriate for **vConTACT2***
2. Creates a gene-to-genome lookup table
3. Runs **vConTACT2**<br>
   *To see available RefSeq reference databases, run* `vcontact2 -h`
4. Runs automated taxonomic predictions using `tax_predict.py`

> **Note on `tax_predict.py`**<br>
> This version of the script was based on the output generated by vConTACT2 (version 0.11.3). It may not work well with other versions of **vConTACT2**.

### Install ClusterONE (if required)

```bash
cd /path/to/conda/envs/vContact2/bin/
wget https://paccanarolab.org/static_content/clusterone/cluster_one-1.0.jar
```

### Run vConTACT2

```bash
#!/bin/bash -e
#SBATCH -A your_project_account
#SBATCH -J 6_Taxonomy_vcontact2
#SBATCH --time 03:00:00
#SBATCH --mem=20GB
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=32
#SBATCH -e 6_Taxonomy_vcontact2.err
#SBATCH -o 6_Taxonomy_vcontact2.out
#SBATCH --profile=task

# Load modules
module purge
module load Prodigal/2.6.3-GCC-9.2.0 Miniconda3 DIAMOND/0.9.32-GCC-9.2.0 MCL/14.137-gimkl-2020a

# Append paths
export PATH="/path/to/conda/envs/vContact2/bin:$PATH"

# Output directories
mkdir -p viral_taxonomy/vConTACT2

# Run prodigal
prodigal -p meta -q \
  -i viral_identification/checkv_vOTUs/vOTUs.checkv_filtered.fna \
  -a viral_taxonomy/vConTACT2/vOTUs.faa 

# Generate gene2genome mapping file
source activate vContact2

vcontact2_gene2genome \
  -p viral_taxonomy/vConTACT2/vOTUs.faa \
  -o viral_taxonomy/vConTACT2/viral_genomes_g2g.csv \
  -s 'Prodigal-FAA'

# Run vcontact2
vcontact2 \
  -t 32 \
  --raw-proteins viral_taxonomy/vConTACT2/vOTUs.faa \
  --rel-mode Diamond \
  --proteins-fp viral_taxonomy/vConTACT2/viral_genomes_g2g.csv \
  --db 'ProkaryoticViralRefSeq211-Merged' \
  --c1-bin /path/to/conda/envs/vContact2/bin/cluster_one-1.0.jar \
  --output-dir viral_taxonomy/vConTACT2/vConTACT2_Results

# Deactivate conda environment
conda deactivate

# Predict taxonomy
module purge
module load Python/3.8.2-gimkl-2020a

tax_predict_vConTACT2.0.11.3.py \
  -i viral_taxonomy/vConTACT2/vConTACT2_Results/genome_by_genome_overview.csv \
  -o viral_taxonomy/vConTACT2/vOTUs_
```
