{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "03ec4473-deaf-4bed-a9f4-cfd682155d3d",
   "metadata": {},
   "source": [
    "# Workflow: GC, AA, and pI\n",
    "\n",
    "Workflow for compiling GC content, amino acid proportions, and isoelectric points of predicted genes and proteins, as used in the study *'Genomic markers show the proliferation of DNA viruses is constrained by adaptation to environmental niche'*.\n",
    "\n",
    "Method: \n",
    "\n",
    "- Per sample:\n",
    "  - Predict genes and protein sequence via DRAMv (prodigal-gv)\n",
    "  - Calculate animo acid proportions per predicted protein\n",
    "  - Generate summaries for AA types including acidic, basic, polar, nonpolar, and charged amino acids (defined as per pepstats (EMBOSS v6.6.0))\n",
    "  - Identify gene GC content\n",
    "  - Output summary table of results\n",
    "  - Also calculate protein isolectric points (pI)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0231cc57-0b21-4f02-89ee-b3b8e9178169",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37196040",
   "metadata": {},
   "source": [
    "## Waiwera estuary data\n",
    "\n",
    "Note re: study design\n",
    "\n",
    "- This was developed based on metagenome assemblies from water and sediment samples at nine sites spanning a salinity gradient in the Waiwera estuary, New Zealand.\n",
    "- Samples were labelled S1-S9 for Filter and sediment samples (e.g. S1.Filter; S1.sediment)\n",
    "- Gene prediction and annotation was conducted via DRAMv for all putative viral contigs identified from each assembly\n",
    "- This workflow uses the following output files generated by DRAMv: `genes.faa`; `final-viral-combined-for-dramv.gbk`; `annotations.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c42f46fa",
   "metadata": {},
   "source": [
    "### Protein AA proportions and gene GC content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46962c70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load python\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "python3\n",
    "\n",
    "### Import required libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import re\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Establish empty variable lists to append results\n",
    "sampleID_list = []\n",
    "contigID_list = []\n",
    "geneID_list = []\n",
    "gene_count_list = []\n",
    "start_list = []\n",
    "end_list = []\n",
    "nt_seq_length_list = []\n",
    "gc_content_list = []\n",
    "seq_aa_list = []\n",
    "aa_seq_length_list = []\n",
    "glu_list = []\n",
    "asp_list = []\n",
    "arg_list = []\n",
    "lys_list = []\n",
    "hist_list = []\n",
    "glycine_list = []\n",
    "serine_list = []\n",
    "tyrosine_list = []\n",
    "cysteine_list = []\n",
    "glutamine_list = []\n",
    "asparagine_list = []\n",
    "threonine_list = []\n",
    "phenylalanine_list = []\n",
    "leucine_list = []\n",
    "tryptophan_list = []\n",
    "proline_list = []\n",
    "isoleucine_list = []\n",
    "methionine_list = []\n",
    "valine_list = []\n",
    "alanine_list = []\n",
    "aa_acidic_list = []\n",
    "aa_basic_list = []\n",
    "aa_polar_list = []\n",
    "aa_nonpolar_list = []\n",
    "aa_charged_list = []\n",
    "\n",
    "# Loop through each sample genes.faa file to extract values for each variable and append to lists\n",
    "# Also calculate propotions for amino  acids and summaries of aa types.\n",
    "for i in range(1, 10):\n",
    "    for j in ['Filter', 'sediment']:\n",
    "        with open('8.DRAMv/persample/S'+str(i)+'.'+j+'/dramv_annotation/genes.faa', 'r') as read_fasta:\n",
    "            for name, seq in SimpleFastaParser(read_fasta):\n",
    "                headers = name.split(' ')\n",
    "                sampleID_list.append('S'+str(i)+'.'+j)\n",
    "                contigID_list.append(re.sub('_\\d+$', '', headers[0].strip()))\n",
    "                geneID_list.append(headers[0].strip())\n",
    "                gene_count_list.append(re.sub(r'.*_(\\d+)$', r'\\1', headers[0].strip()))\n",
    "                seq_aa_list.append(seq)\n",
    "                aa_seq_length = len(seq)\n",
    "                aa_seq_length_list.append(aa_seq_length)\n",
    "                glu_list.append(seq.count('E')/aa_seq_length)\n",
    "                asp_list.append(seq.count('D')/aa_seq_length)\n",
    "                arg_list.append(seq.count('R')/aa_seq_length)\n",
    "                lys_list.append(seq.count('K')/aa_seq_length)\n",
    "                hist_list.append(seq.count('H')/aa_seq_length)\n",
    "                glycine_list.append(seq.count('G')/aa_seq_length)\n",
    "                serine_list.append(seq.count('S')/aa_seq_length)\n",
    "                tyrosine_list.append(seq.count('Y')/aa_seq_length)\n",
    "                cysteine_list.append(seq.count('C')/aa_seq_length)\n",
    "                glutamine_list.append(seq.count('Q')/aa_seq_length)\n",
    "                asparagine_list.append(seq.count('N')/aa_seq_length)\n",
    "                threonine_list.append(seq.count('T')/aa_seq_length)\n",
    "                phenylalanine_list.append(seq.count('F')/aa_seq_length)\n",
    "                leucine_list.append(seq.count('L')/aa_seq_length)\n",
    "                tryptophan_list.append(seq.count('W')/aa_seq_length)\n",
    "                proline_list.append(seq.count('P')/aa_seq_length)\n",
    "                isoleucine_list.append(seq.count('I')/aa_seq_length)\n",
    "                methionine_list.append(seq.count('M')/aa_seq_length)\n",
    "                valine_list.append(seq.count('V')/aa_seq_length)\n",
    "                alanine_list.append(seq.count('A')/aa_seq_length)\n",
    "                aa_acidic_list.append((seq.count('E')/aa_seq_length)+(seq.count('D')/aa_seq_length))\n",
    "                aa_basic_list.append((seq.count('R')/aa_seq_length)+(seq.count('K')/aa_seq_length)+(seq.count('H')/aa_seq_length))\n",
    "                aa_polar_list.append((seq.count('D')/aa_seq_length)+(seq.count('E')/aa_seq_length)+(seq.count('H')/aa_seq_length)+(seq.count('K')/aa_seq_length)+(seq.count('N')/aa_seq_length)+(seq.count('Q')/aa_seq_length)+(seq.count('R')/aa_seq_length)+(seq.count('S')/aa_seq_length)+(seq.count('T')/aa_seq_length)+(seq.count('Z')/aa_seq_length))\n",
    "                aa_nonpolar_list.append((seq.count('A')/aa_seq_length)+(seq.count('C')/aa_seq_length)+(seq.count('F')/aa_seq_length)+(seq.count('G')/aa_seq_length)+(seq.count('I')/aa_seq_length)+(seq.count('L')/aa_seq_length)+(seq.count('M')/aa_seq_length)+(seq.count('P')/aa_seq_length)+(seq.count('V')/aa_seq_length)+(seq.count('W')/aa_seq_length)+(seq.count('Y')/aa_seq_length))\n",
    "                aa_charged_list.append((seq.count('B')/aa_seq_length)+(seq.count('D')/aa_seq_length)+(seq.count('E')/aa_seq_length)+(seq.count('H')/aa_seq_length)+(seq.count('K')/aa_seq_length)+(seq.count('R')/aa_seq_length)+(seq.count('Z')/aa_seq_length))\n",
    "\n",
    "df = pd.DataFrame({'sampleID': sampleID_list, \n",
    "                   'contigID': contigID_list, \n",
    "                   'geneID': geneID_list, \n",
    "                   'gene_count': gene_count_list,\n",
    "                   'aa_seq': seq_aa_list,\n",
    "                   'aa_seq_length': aa_seq_length_list,\n",
    "                   'aa_prop_Glutamic_acid': glu_list,\n",
    "                   'aa_prop_Aspartic_acid': asp_list,\n",
    "                   'aa_prop_Arginine': arg_list,\n",
    "                   'aa_prop_Lysine': lys_list,\n",
    "                   'aa_prop_Histidine': hist_list,\n",
    "                   'aa_prop_Glycine': glycine_list,\n",
    "                   'aa_prop_Serine': serine_list,\n",
    "                   'aa_prop_Tyrosine': tyrosine_list,\n",
    "                   'aa_prop_Cysteine': cysteine_list,\n",
    "                   'aa_prop_Glutamine': glutamine_list,\n",
    "                   'aa_prop_Asparagine': asparagine_list,\n",
    "                   'aa_prop_Threonine': threonine_list,\n",
    "                   'aa_prop_Phenylalanine': phenylalanine_list,\n",
    "                   'aa_prop_Leucine': leucine_list,\n",
    "                   'aa_prop_Tryptophan': tryptophan_list,\n",
    "                   'aa_prop_Proline': proline_list,\n",
    "                   'aa_prop_Isoleucine': isoleucine_list,\n",
    "                   'aa_prop_Methionine': methionine_list,\n",
    "                   'aa_prop_Valine': valine_list,\n",
    "                   'aa_prop_Alanine': alanine_list,\n",
    "                   'aa_prop_Acidic': aa_acidic_list,\n",
    "                   'aa_prop_Basic': aa_basic_list,\n",
    "                   'aa_prop_Polar': aa_polar_list,\n",
    "                   'aa_prop_Nonpolar': aa_nonpolar_list,\n",
    "                   'aa_prop_charged': aa_charged_list\n",
    "                  })\n",
    "\n",
    "\n",
    "# Pull in annotations and gbk files to make df of other fields of interest (e.g. start, end, seq length, gc content)\n",
    "annot_dfs = []\n",
    "for i in range(1, 10):\n",
    "    for j in ['Filter', 'sediment']:\n",
    "        gbk_seq_ids = []\n",
    "        gbk_gene_ids = []\n",
    "        gbk_gene_gc = []\n",
    "        with open('8.DRAMv/persample/S'+str(i)+'.'+j+'/dramv_annotation/genbank/final-viral-combined-for-dramv.gbk', 'r') as genbank_infile:\n",
    "            for seq_record in SeqIO.parse(genbank_infile, \"genbank\"):\n",
    "                for gene in seq_record.features:\n",
    "                    if gene.type == 'CDS':\n",
    "                        gbk_seq_ids.append(seq_record.id)\n",
    "                        gbk_gene_ids.append(gene.qualifiers['gene'][0])\n",
    "                        gbk_gene_gc.append(gene.qualifiers['gc_cont'][0])\n",
    "        annot_df = pd.read_csv('8.DRAMv/persample/S'+str(i)+'.'+j+'/dramv_annotation/annotations.tsv', sep='\\t').rename(columns={'Unnamed: 0':'geneID', 'start_position': 'start','end_position': 'end'})\n",
    "        annot_df = annot_df[['geneID', 'start', 'end']+[col for col in annot_df if 'vogdb' in col]]\n",
    "        annot_df = pd.merge(\n",
    "            pd.DataFrame({'contigID': gbk_seq_ids, 'geneID': gbk_gene_ids, 'gene_gc': gbk_gene_gc}),\n",
    "            annot_df,\n",
    "            how = 'outer', \n",
    "            on = 'geneID')\n",
    "        annot_df['sampleID'] = 'S'+str(i)+'.'+j\n",
    "        annot_dfs.append(annot_df)\n",
    "\n",
    "annot_df_full = pd.concat(annot_dfs).reset_index(drop=True)\n",
    "\n",
    "full_df = pd.merge(df, annot_df_full, how='left', on=['sampleID','contigID', 'geneID'])\n",
    "full_df = full_df.rename(columns={'start': 'nt_start', 'end': 'nt_end', 'gene_gc': 'gc_content'})\n",
    "full_df['nt_seq_length'] = full_df['nt_end'] - full_df['nt_start']\n",
    "\n",
    "# write out summary table\n",
    "full_df.to_csv('AA_and_GC.summary_table.tsv', sep='\\t', index=False)\n",
    "\n",
    "quit()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ff2d063",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "### Protein isoelectric point (pI)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "499685b1",
   "metadata": {},
   "source": [
    "Run pepstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba0ab815",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# Set up working directories\n",
    "module load EMBOSS/6.6.0-gimkl-2020a\n",
    "\n",
    "for i in {1..9}; do\n",
    "    for j in Filter sediment; do\n",
    "        echo \"Sample: S${i}.${j}\"\n",
    "        pepstats \"8.DRAMv/persample/S${i}.${j}/dramv_annotation/genes.faa\" -outfile \"S${i}.${j}.pepstats\"\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21d6d293",
   "metadata": {},
   "source": [
    "Summarise pepstats pI results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96462546",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load python\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "python3\n",
    "\n",
    "### Import required libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import re\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from itertools import groupby, chain\n",
    "\n",
    "def get_gene_pepstats(file):\n",
    "    with open(file) as f:\n",
    "        grps = groupby(f, key=lambda x: x.lstrip().startswith(\"PEPSTATS of\"))\n",
    "        for k, v in grps:\n",
    "            if k:\n",
    "                yield chain([next(v)], (next(grps)[1]))  \n",
    "\n",
    "iep_dfs = []\n",
    "for i in range(1, 10):\n",
    "    for j in ['Filter', 'sediment']:\n",
    "        sampleID = \"S\"+str(i)+\".\"+j\n",
    "        iep_dict = {}\n",
    "        for gene_pepstats in get_gene_pepstats(sampleID+\".pepstats\"):\n",
    "            gene_pepstats_tmp = (list(gene_pepstats))\n",
    "            geneID = [x for x in gene_pepstats_tmp if 'PEPSTATS of' in x][0].split(' ')[2]\n",
    "            iso_point = [x for x in gene_pepstats_tmp if 'Isoelectric Point' in x][0].split(' ')[3].rstrip()\n",
    "            iep_dict[geneID] = iso_point\n",
    "        iep_df_tmp = pd.DataFrame(iep_dict.items(), columns=['geneID', 'isoelectric_point'])\n",
    "        iep_df_tmp['sampleID'] = sampleID\n",
    "        iep_dfs.append(iep_df_tmp)\n",
    "\n",
    "iep_df = pd.concat(iep_dfs).reset_index(drop=True)\n",
    "iep_df.to_csv('pepstats.summaryTable.tsv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8096fbf6",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2c2cf5e",
   "metadata": {},
   "source": [
    "## IMG/VR high-qualtiy viruses\n",
    "\n",
    "Note:\n",
    "\n",
    "- Gene prediction for all high-quality viruses in the IMG/VR database (v7.1) was conducted via prodigal-gv\n",
    "- Gene annotation was conducted via hmmsearch of predicted proteins from the IMG/VR dataset against proteins in the viral orthologous groups (VOG) database\n",
    "- This workflow uses the following files generated by prodigal-gv and hmmsearch: `prodigal_gv.faa`; `final-viral-combined-for-dramv.gbk`; `annotations.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c46f22dd",
   "metadata": {},
   "source": [
    "### Protein AA proportions and gene GC content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c647f62b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load python\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "python3\n",
    "\n",
    "### Import required libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import re\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Establish empty variable lists to append results\n",
    "sampleID_list = []\n",
    "contigID_list = []\n",
    "geneID_list = []\n",
    "gene_count_list = []\n",
    "start_list = []\n",
    "end_list = []\n",
    "nt_seq_length_list = []\n",
    "gc_content_list = []\n",
    "seq_aa_list = []\n",
    "aa_seq_length_list = []\n",
    "glu_list = []\n",
    "asp_list = []\n",
    "arg_list = []\n",
    "lys_list = []\n",
    "hist_list = []\n",
    "glycine_list = []\n",
    "serine_list = []\n",
    "tyrosine_list = []\n",
    "cysteine_list = []\n",
    "glutamine_list = []\n",
    "asparagine_list = []\n",
    "threonine_list = []\n",
    "phenylalanine_list = []\n",
    "leucine_list = []\n",
    "tryptophan_list = []\n",
    "proline_list = []\n",
    "isoleucine_list = []\n",
    "methionine_list = []\n",
    "valine_list = []\n",
    "alanine_list = []\n",
    "aa_acidic_list = []\n",
    "aa_basic_list = []\n",
    "aa_polar_list = []\n",
    "aa_nonpolar_list = []\n",
    "aa_charged_list = []\n",
    "\n",
    "# prodigal.faa file to extract values for each variable and append to lists\n",
    "# Also calculate propotions for amino  acids\n",
    "with open('img_vr_All_HighQuality/prodigal_gv/prodigal_gv.faa', 'r') as read_fasta:\n",
    "    for name, seq in SimpleFastaParser(read_fasta):\n",
    "        if 'IMGVR' in name:\n",
    "            headers = name.split('#')\n",
    "            sampleID_list.append(re.sub('_UViG.*', '', headers[0].strip()))\n",
    "            contigID_list.append(re.sub('_\\d+$', '', headers[0].strip()))\n",
    "            geneID_list.append(headers[0].strip())\n",
    "            gene_count_list.append(re.sub(r'.*_(\\d+)$', r'\\1', headers[0].strip()))\n",
    "            start_list.append(float(headers[1].strip()))\n",
    "            end_list.append(float(headers[2].strip()))\n",
    "            nt_seq_length_list.append(float(headers[2].strip())-float(headers[1].strip()))\n",
    "            gc_content_list.append(float(headers[-1].split('gc_cont=')[-1].strip()))\n",
    "            seq_aa_list.append(seq)\n",
    "            aa_seq_length = len(seq)\n",
    "            aa_seq_length_list.append(aa_seq_length)\n",
    "            glu_list.append(seq.count('E')/aa_seq_length)\n",
    "            asp_list.append(seq.count('D')/aa_seq_length)\n",
    "            arg_list.append(seq.count('R')/aa_seq_length)\n",
    "            lys_list.append(seq.count('K')/aa_seq_length)\n",
    "            hist_list.append(seq.count('H')/aa_seq_length)\n",
    "            glycine_list.append(seq.count('G')/aa_seq_length)\n",
    "            serine_list.append(seq.count('S')/aa_seq_length)\n",
    "            tyrosine_list.append(seq.count('Y')/aa_seq_length)\n",
    "            cysteine_list.append(seq.count('C')/aa_seq_length)\n",
    "            glutamine_list.append(seq.count('Q')/aa_seq_length)\n",
    "            asparagine_list.append(seq.count('N')/aa_seq_length)\n",
    "            threonine_list.append(seq.count('T')/aa_seq_length)\n",
    "            phenylalanine_list.append(seq.count('F')/aa_seq_length)\n",
    "            leucine_list.append(seq.count('L')/aa_seq_length)\n",
    "            tryptophan_list.append(seq.count('W')/aa_seq_length)\n",
    "            proline_list.append(seq.count('P')/aa_seq_length)\n",
    "            isoleucine_list.append(seq.count('I')/aa_seq_length)\n",
    "            methionine_list.append(seq.count('M')/aa_seq_length)\n",
    "            valine_list.append(seq.count('V')/aa_seq_length)\n",
    "            alanine_list.append(seq.count('A')/aa_seq_length)\n",
    "            aa_acidic_list.append((seq.count('E')/aa_seq_length)+(seq.count('D')/aa_seq_length))\n",
    "            aa_basic_list.append((seq.count('R')/aa_seq_length)+(seq.count('K')/aa_seq_length)+(seq.count('H')/aa_seq_length))\n",
    "            aa_polar_list.append((seq.count('D')/aa_seq_length)+(seq.count('E')/aa_seq_length)+(seq.count('H')/aa_seq_length)+(seq.count('K')/aa_seq_length)+(seq.count('N')/aa_seq_length)+(seq.count('Q')/aa_seq_length)+(seq.count('R')/aa_seq_length)+(seq.count('S')/aa_seq_length)+(seq.count('T')/aa_seq_length)+(seq.count('Z')/aa_seq_length))\n",
    "            aa_nonpolar_list.append((seq.count('A')/aa_seq_length)+(seq.count('C')/aa_seq_length)+(seq.count('F')/aa_seq_length)+(seq.count('G')/aa_seq_length)+(seq.count('I')/aa_seq_length)+(seq.count('L')/aa_seq_length)+(seq.count('M')/aa_seq_length)+(seq.count('P')/aa_seq_length)+(seq.count('V')/aa_seq_length)+(seq.count('W')/aa_seq_length)+(seq.count('Y')/aa_seq_length))\n",
    "            aa_charged_list.append((seq.count('B')/aa_seq_length)+(seq.count('D')/aa_seq_length)+(seq.count('E')/aa_seq_length)+(seq.count('H')/aa_seq_length)+(seq.count('K')/aa_seq_length)+(seq.count('R')/aa_seq_length)+(seq.count('Z')/aa_seq_length))\n",
    "\n",
    "df = pd.DataFrame({'sampleID': sampleID_list, \n",
    "                   'contigID': contigID_list, \n",
    "                   'geneID': geneID_list, \n",
    "                   'aa_prop_Lysine': lys_list,\n",
    "                   'aa_prop_Histidine': hist_list,\n",
    "                   'aa_prop_Glycine': glycine_list,\n",
    "                   'aa_prop_Serine': serine_list,\n",
    "                   'aa_prop_Tyrosine': tyrosine_list,\n",
    "                   'aa_prop_Cysteine': cysteine_list,\n",
    "                   'aa_prop_Glutamine': glutamine_list,\n",
    "                   'aa_prop_Asparagine': asparagine_list,\n",
    "                   'aa_prop_Threonine': threonine_list,\n",
    "                   'aa_prop_Phenylalanine': phenylalanine_list,\n",
    "                   'aa_prop_Leucine': leucine_list,\n",
    "                   'aa_prop_Tryptophan': tryptophan_list,\n",
    "                   'aa_prop_Proline': proline_list,\n",
    "                   'aa_prop_Isoleucine': isoleucine_list,\n",
    "                   'aa_prop_Methionine': methionine_list,\n",
    "                   'aa_prop_Valine': valine_list,\n",
    "                   'aa_prop_Alanine': alanine_list,\n",
    "                   'aa_prop_Acidic': aa_acidic_list,\n",
    "                   'gene_count': gene_count_list,\n",
    "                   'nt_start': start_list, \n",
    "                   'nt_end': end_list, \n",
    "                   'nt_seq_length': nt_seq_length_list,\n",
    "                   'gc_content': gc_content_list, \n",
    "                   'aa_seq': seq_aa_list,\n",
    "                   'aa_seq_length': aa_seq_length_list,\n",
    "                   'aa_prop_Glutamic_acid': glu_list,\n",
    "                   'aa_prop_Aspartic_acid': asp_list,\n",
    "                   'aa_prop_Arginine': arg_list,\n",
    "                   'aa_prop_Basic': aa_basic_list,\n",
    "                   'aa_prop_Polar': aa_polar_list,\n",
    "                   'aa_prop_Nonpolar': aa_nonpolar_list,\n",
    "                   'aa_prop_charged': aa_charged_list\n",
    "                  })\n",
    "\n",
    "# write out summary table\n",
    "df_filt = df[['sampleID','contigID','geneID','gc_content']+[col for col in df.columns if 'aa_prop' in col]]\n",
    "df_filt.to_csv('img_vr_All_HighQuality.AA_and_GC.summary_table.tsv', sep='\\t', index=False)\n",
    "\n",
    "quit()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2e0fb2f",
   "metadata": {},
   "source": [
    "## Protein isoelectric point (pI)\n",
    "\n",
    "- note: pepstats trims IDs at '|' character, so need to edit these"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cce88179",
   "metadata": {},
   "source": [
    "Run pepstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55e12461",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# Replace '|' characters with '__'\n",
    "cp img_vr_All_HighQuality/prodigal_gv/prodigal_gv.faa img_vr_All_HighQuality/\n",
    "sed -i -e 's/|/__/g' img_vr_All_HighQuality/prodigal_gv.faa\n",
    "\n",
    "module load EMBOSS/6.6.0-gimkl-2020a\n",
    "\n",
    "pepstats \"img_vr_All_HighQuality/prodigal_gv.faa\" -outfile \"img_vr_All_HighQuality/img_vr_HighQuality.pepstats\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62bb9298",
   "metadata": {},
   "source": [
    "Summarise pepstats pI results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5395d03c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load python\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "python3\n",
    "\n",
    "### Import required libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import re\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from itertools import groupby, chain\n",
    "\n",
    "def get_gene_pepstats(file):\n",
    "    with open(file) as f:\n",
    "        grps = groupby(f, key=lambda x: x.lstrip().startswith(\"PEPSTATS of\"))\n",
    "        for k, v in grps:\n",
    "            if k:\n",
    "                yield chain([next(v)], (next(grps)[1]))  \n",
    "\n",
    "iep_dict = {}\n",
    "for gene_pepstats in get_gene_pepstats(\"img_vr_All_HighQuality/img_vr_HighQuality.pepstats\"):\n",
    "    gene_pepstats_tmp = (list(gene_pepstats))\n",
    "    geneID = [x for x in gene_pepstats_tmp if 'PEPSTATS of' in x][0].split(' ')[2].replace('__', '|')\n",
    "    iso_point = [x for x in gene_pepstats_tmp if 'Isoelectric Point' in x][0].split(' ')[3].rstrip()\n",
    "    iep_dict[geneID] = iso_point\n",
    "\n",
    "iep_df = pd.DataFrame(iep_dict.items(), columns=['geneID', 'isoelectric_point'])\n",
    "iep_df['sampleID'] = 'imgvr'\n",
    "\n",
    "iep_df.to_csv('img_vr_All_HighQuality/img_vr_HighQuality.pepstats.summaryTable.tsv', sep='\\t', index=False)\n",
    "\n",
    "quit()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df214f03",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e568855b",
   "metadata": {},
   "source": [
    "***"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
