# Getting started

For this workflow, we are using a mixture of previously generated files as well as some new data. In addition to the previous software, we are going to use:

1. [**BBMap**](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbmap-guide/) v38.95
1. [**samtools**](https://github.com/samtools/samtools) v1.13
1. [**prodigal**](https://github.com/hyattpd/Prodigal) v2.6.3
1. [**subread**](http://subread.sourceforge.net/) v2.0.0

The above softwares are all available on NeSI as modules.

Prior to starting this workflow, we will need to QC and trim the RNA sequence files as we have done with metagenomic reads (remove adapter sequences, trim poor quality regions).

----

#### Mapping RNA sequences to the assembly

Like with the DNA read mapping, we can map our RNA data to the assembly using **BBMap**, **bowtie**, or **bowtie2**. This is at your discretion, although we will use **BBMap** here. We then sort and compress the *sam* alignment using **samtools**.

```bash
bbmap.sh ref=contigs.fna

bbmap.sh t=30 ref=contigs.fna in1=SampleA_R1.rna.fq in2=SampleA_R2.rna.fq out=SampleA.rna.sam
bbmap.sh t=30 ref=contigs.fna in1=SampleB_R1.rna.fq in2=SampleB_R2.rna.fq out=SampleB.rna.sam

samtools sort -@ 10 -o SampleA.rna.bam SampleA.rna.sam
samtools sort -@ 10 -o SampleB.rna.bam SampleB.rna.sam
```

----

#### Creating feature table for featureCounts

We now need to create a table mapping out the coordinates of contigs which correspond to genes. In this example, we are only going to look at protein-coding sequences identified by **prodigal**, but it is informative to include tRNA and rRNA annotations as well.

Creating the table can be done easily with **python**, specifically the **pandas** library. Please note that this example is written to work under **python** version 3 and will not work for version 2.

```python
import pandas as pd
import re

# Import the prodigal file, and extract sequence names/metadata
prodigal_headers = [ line.replace('>', '') for line in open('genome.prod.faa', 'r') if '>' in line ]

# Define a function for splitting the metadata into row-wise dictionaries
def prodigal_to_dict(file_line):
    gene_name, start, stop, orientation, *rest = file_line.split('#')
    return { 'GeneID': gene_name, 
             'Chr': re.sub('_[\d+]$', '', gene_name),
             'Start': start.strip(),
             'End': stop.strip(),
             'Strand': '+' if int(orientation) == 1 else '-' }

# Convert the prodigal lines into a DataFrame
prodigal_buffer = [ prodigal_to_dict(header) for header in prodigal_headers ]
prodigal_df = pd.DataFrame(prodigal_buffer)

# Order the columns, and save the file
prodigal_df = prodigal_df[ ['GeneID', 'Chr', 'Start', 'End', 'Strand' ] ]
prodigal_df.to_csv('gene_coords.txt', sep='\t', index=False)
```

----

#### Distributing the RNA counts over gene regions

This can be done with the **featureCounts** program from the **subread** library.

```bash
featureCounts -p -T 20 -t exon -F SAF -a gene_coords.txt -o gene_counts.txt *.rna.bam
```

The results of this are stored to the file `gene_counts.txt`, which can then be imported into a program like **R** to identify the differentially expressed genes.

---

#### Obtaining read mapping statistics

Statistics from the read mapping process such as number of reads and read lengths are important for differential expression analyses. Here, we can use the stat function from SAMtools then subsetting the relevent lines from the output file. The outputs from the following lines will give us the library size as total number of reads and number of mapped reads, as well as the distribution of read lengths, in each sample.

```bash
# Obtain read mapping statistics
for file in *.rna.sam; do # This can also be used with BAM files
  sample=$(basename ${file} .sam)
  samtools stat -@ 1 ${file} > ${sample}.stat
done

# Subset relevent information from .stat files
printf "sample\ttotal_reads\tmapped_reads\n" > lib_size.txt
printf "sample\tread_length\tfrequency\n" > read_length.txt
for file in *.stat; do
  sample=$(basename ${file} .stat)
  # For library size
  total_reads=$(grep '^SN' ${j} | cut -f 2- | grep '^sequences:' | cut -f 2)
  mapped_reads=$(grep '^SN' ${j} | cut -f 2- | grep '^reads mapped:' | cut -f 2)
  printf "%s\t%d\t%d\n" ${sample} ${total_reads} ${mapped_reads} >> lib_size.txt
  # For read length
  grep '^RL' ${j} | cut -f 2- | \ 
  awk -v sample="${sample}" '{print sample"\t"$0}' >> read_length.txt
done
```

---

## Next steps

Now that we have the expressed gene count matrix and associated statistics, there are a few analyses we can do. Most often, the next step would be to perform differential expression analyses. This is usually done in R aided by packages, some of the most popular are: 
  - [**edgeR**](https://bioconductor.org/packages/release/bioc/html/edgeR.html) by [Robinson, McCarthy & Smyth (2010)](https://doi.org/10.1093/bioinformatics/btp616)
  - [**DESeq2**](http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html) by [Love, Huber & Anders (2014)](https://doi.org/10.1186/s13059-014-0550-8)
  - [**limma**](https://bioinf.wehi.edu.au/limma/) by [Ritchie et al. (2015)](https://doi.org/10.1093/nar/gkv007)

For more in-depth workflows on differential expression analyses, please consult the [RNA-seq workshop materials](https://github.com/GenomicsAotearoa/RNA-seq-workshop) where both theory and practical usage is discussed.

Network analysis are also useful to find clusters of correlated genes. A tool we have used in the Handley lab (see [Sze et al., 2021](https://doi.org/10.1038/s41396-020-0676-5)) is [**WGCNA**](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/) ([Langfelder & Hovarth, 2008](https://doi.org/10.1186/1471-2105-9-559)). Please note that this package has some factors that are important for consideration:
  - Knowledge of how to use/run R scripts in an HPC environment is necessary if the dataset is large. Large datasets will lead to the generation of very large matrices that require much more memory to run. 
  - Computation time is highly dependent on the BLAS library that R was compiled against.
  - Not all sections of the analysis can be multi-threaded.

---

