# Manual bin refinement

Once the initial dereplication has been performed, it is helpful to use a more robust clustering algorithm to view your bins and make sure they they represent distinct clusters of genomic material. For a quick view, you can use the program **VizBin** which provides a great set of tools to perform a visualisation using the *t-Distributed Stochastic Neighbor Embedding* ([t-SNE](https://lvdmaaten.github.io/publications/papers/JMLR_2014.pdf)) algorithm. For a simple approach, you can load your data straight into the **VizBin** program like so:

```bash
cat dRep_output/dereplicated_genomes/* > all_bins.fna

echo "label" > all_bins.ann.txt

for i in `dRep_output/dereplicated_genomes/*`;
do
    for j in `grep ">" ${i}`;
    do
        echo ${i} >> all_bins.ann.txt
    done
done
```

You can then execute **VizBin** as per the manual, and use the file *all_bins.fna* as the input, and *all_bins.ann.txt* as the annotation file for colouring points.This should give you a good first pass at resolving your bins. However, there is one big issue with **VizBin** - it does not allow for multiple coverage values per contig.

VizBin is also helpful for identifying problematic contigs. You can find an example of this use [here](https://genomicsaotearoa.github.io/metagenomics_summer_school/day2/ex9_refining_bins/#optional-refine-and-filter-problematic-contigs-from-bins).

# Manual bin refinement with differential coverage (*experimental*)
If you have used differential coverage as part of your binning workflow, and want to make use of this data, we have a set of scripts for performing the *t-SNE* transformation across data with multiple coverage values, and allowing the user to up- or down-weight the importance of coverage during the process.

> ***Note: This is as experimental part of our workflow. Use at your own risk.***

## Variable-coverage *t-SNE* projection
A more advanced way of performing the *t-SNE* projection over a data set is to consider the following factors:

1. Splitting long contigs into smaller pieces, similar to the way **CONCOCT** performs binning. This is helpful for increasing the density of clustering in the *t-SNE*.
1. Using coverage from multiple samples to add dimensionality to the binning process.
1. Varying the relative importance of coverage compared with k-mer weighting.
1. Scaling the k-mer and coverage data prior to performing the *t-SNE* projection.

There are only two scripts we use here, but there are a few parameters that need to be considered for each one.

```bash
# Chop up the fragments in each bin, then pool them.
# Simulatenously create the annotation file.

echo "label" > all_bins.ann.10k.txt

for i in `dRep_output/dereplicated_genomes/*`;
do
    cut_up_fasta.py ${i} -c 10000 -o 0 --merge_last > ${i}.chop
    for j in `grep ">" ${i}.chop`;
    do
        echo ${i} >> all_bins.ann.10k.txt
    done
    cat ${i}.chop >> all_bins.10k.fa
    rm ${i}.chop
done

# Compute a k-mer distribution file over the 
compute_kmer_profile.py -t 10 -n yeojohnson -k 4 -c contigs.10k.txt all_bins.10k.fa
```

The **compute_kmer_profile.py** calculates the k-mer frequency profile for each contig in the input file, and produces an output table with the form '\<*input file*\>.tsv'. There are several parameters you can tune here:

1. **-k**: K-mer size. The default is to use tetranucleotide (k = 4) frequencies, but this can be increased if desired. For difficult to resolve bins, increasing to 5 can improve resolution.
1. **-n**: Normalisation. Peform column-wise (feature-wise) normalisation on the resulting table. The default option is to perform unit variance scaling. Other options are Yeo-Johnson or none.
1. **-c**: Coverage table. Optional, but you can include coverage information as a tab-delimited table where the first column is the contig names, and subsequent columns are coverage values for a particular sample. Sample names will be replaced with names 'Cov1', 'Cov2' etc., and will be scaled the same way as kmers. When adding coverage information, kmer and coverage information is matched using an inner join, so only entries that exist in both the contigs file and coverage table are reported. This means that you can just use the original mapping file, and all contigs that were not included in the final bins will be discarded automatically.
1. **-t**: Number of threads to use.

The actual *t-SNE* projection can be performed using the script **project_tsne.py.py**.

```bash
project_tsne.py -o all_bins.10k.uniform.tsne all_bins.10k.tsv
project_tsne.py -w 0.5 -o all_bins.10k.weighted.tsne all_bins.10k.tsv
```

This script takes the output of the **compute_kmer_profile.py** script as a positional argument, and allows you to specify an output name with the parameter **-o**. The important parameter to note in this script is **-w**. This is the weighting factor for the coverage columns.

Weighting of coverage values is an important consideration in performing this transformation, as it is almost a guarantee that you have more observed kmers in your data than coverage sites. For example, if you have an assembly from a pair of sites as in this workflow, then as a percentage of features considered in clustering, then for k=4 your coverage will only account for about 7.7% of the signal in your data.

<img src="https://latex.codecogs.com/svg.latex?\frac{N.\text{&space;}coverage\text{&space;}features}{4!&space;&plus;&space;N.\text{&space;}coverage\text{&space;}features}&space;=&space;\frac{2}{24&space;&plus;&space;2}&space;\approx&space;0.0769" title="\frac{N. coverage features}{4! + N. coverage features} = \frac{2}{24 + 2} \approx 0.0769" title="Column weighting example" />

We can therefore use the **-w** flag to increase the importance of coverage and make abundance a more important feature. To use this parameter, specify a value between 0 and 1 that is the proportion of weighting you want to attribute to coverage. This will then be divided amongst the number of coverage columns, and the remainder divided amongst each of the kmers. I.e.

<img src="https://latex.codecogs.com/svg.latex?Coverage\text{&space;}weighting&space;=&space;\frac{0.5}{N.\text{&space;}coverage\text{&space;}features}&space;=&space;\frac{0.5}{2}&space;=&space;0.25" title="Coverage weighting" />

<img src="https://latex.codecogs.com/svg.latex?Kmer\text{&space;}weighting&space;=&space;\frac{1&space;-&space;0.5}{4!}&space;=&space;\frac{0.5}{24}&space;\approx&space;0.0208" title="Kmer weighting" />

We usually perform clustering using the values in the example below, and then only look at other options if there are problematic regions in the plot.




